metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/MerkleTree.scala

```diff
@@ -58,9 +61,124 @@ object MerkleTree {
       // Position of the leaf in the lowest level.
       leafIndex: Int,
       // Number of leaves in the lowest level.
+      // TODO: remove? not used when verifying the proof
       leafCount: Int,
       // Hashes of the "other" side of the tree, level by level,
       // starting from the lowest up to the highest.
-      siblingPath: IndexedSeq[MerkleTree.Hash]
+      siblingPath: IndexedSeq[Hash]
   )
+
+  def build(
+      elems: Iterable[Hash],
+      hashFn: (Hash, Hash) => Hash = concatThenKec256
+  ): MerkleTree = {
+    @tailrec
+    def buildTree(nodes: Seq[MerkleTree]): Seq[MerkleTree] = {
+      if (nodes.size <= 1)
+        nodes
+      else {
+        val paired = nodes.grouped(2).toSeq.map {
+          case Seq(a, b) =>
+            MerkleTree(hashFn(a.hash, b.hash), Some(a), Some(b))
+          case Seq(a) =>
+            // if the element has no pair we hash it with itself
+            MerkleTree(hashFn(a.hash, a.hash), Some(a))
```

Comment:
I think all those options are viable. I initially just hashed the hash, but I ended up with self-concat thinking it would be least controversial, as that's what bitcoin does. The only thing I can think of is that the approaches you mention do reveal a bit more about the tree structure in the proof. Can't see how that could lead to any vulnerabilities.

Reply:
Thanks, didn't know this was how bitcoin did it; that's good enough for me :)